openwisp_controller/connection/base/models.py

```diff
@@ -110,27 +110,35 @@ def __str__(self):
 
     def save(self, *args, **kwargs):
         super().save(*args, **kwargs)
-        self.auto_add_to_devices()
+        if self.auto_add:
+            transaction.on_commit(
+                lambda: auto_add_credentials_to_devices.delay(
+                    credential_id=self.pk, organization_id=self.organization_id
+                )
+            )
 
-    def auto_add_to_devices(self):
+    @classmethod
+    def auto_add_to_devices(cls, credential_id, organization_id):
         """
         When ``auto_add`` is ``True``, adds the credentials
         to each relevant ``Device`` and ``DeviceConnection`` objects
         """
-        if not self.auto_add:
-            return
-        device_model = load_model('config', 'Device')
-        devices = device_model.objects.exclude(config=None)
-        org = self.organization
-        if org:
-            devices = devices.filter(organization=org)
+        DeviceConnection = load_model('connection', 'DeviceConnection')
+        Device = load_model('config', 'Device')
+
+        devices = Device.objects.exclude(config=None)
+        if organization_id:
+            devices = devices.filter(organization_id=organization_id)
         # exclude devices which have been already added
-        devices = devices.exclude(deviceconnection__credentials=self)
+        devices = devices.exclude(deviceconnection__credentials_id=credential_id)
+        device_connections = []
         for device in devices:
-            DeviceConnection = load_model('connection', 'DeviceConnection')
```

Comment:
I have a doubt with the current implementation.

Say the credentials need to be addded on 1 million devices.
What happens here?
The 1 million device instances are loaded in memory, validated and only at the end saved.

But this may eat up the memory of the server.

So I am not sure this is the best approach.
Using bulk_create is surely faster than creating credentials one by one, but we can't do this if the number is huge.

Please take a look at [iterator()](https://docs.djangoproject.com/en/3.2/ref/models/querysets/), which except for MySQL, avoids to iterate over a queryset in batches of a specific size (by default 2000).

We could probably do something like:

```python
device_connections = []
chunk_size = 1000

for device in devices.iterator(chunk_size=chunk_size):
    # ...
    device_connections.append(conn)

    # send create query when chunk_size is reached
    # and reset the device_connections list
    if len(device_connections) >= chunk_size:
        DeviceConnection.objects.bulk_create(device_connections)
        device_connections = []
```

Please try and let me know.